src/neuralnet/project.cpp

```diff
@@ -0,0 +1,210 @@
+    if (m_name == "World_Community_Grid") {
+        return "https://www.worldcommunitygrid.org/";
+    }
+
+    return BaseUrl();
+}
+
+std::string Project::StatsUrl(const std::string& type) const
+{
+    if (type.empty()) {
+        return BaseUrl() + "stats/";
+    }
+
+    return BaseUrl() + "stats/" + type + ".gz";
+}
+
+// -----------------------------------------------------------------------------
+// Class: WhitelistSnapshot
+// -----------------------------------------------------------------------------
+
+WhitelistSnapshot::WhitelistSnapshot(const ProjectListPtr projects)
+    : m_projects(projects)
+{
+}
+
+WhitelistSnapshot::const_iterator WhitelistSnapshot::begin() const
+{
+    return m_projects->begin();
+}
+
+WhitelistSnapshot::const_iterator WhitelistSnapshot::end() const
+{
+    return m_projects->end();
+}
+
+WhitelistSnapshot::size_type WhitelistSnapshot::size() const
+{
+    return m_projects->size();
+}
+
+bool WhitelistSnapshot::Populated() const
+{
+    return !m_projects->empty();
+}
+
+bool WhitelistSnapshot::Contains(const std::string& name) const
+{
+    if (name.empty()) {
+        return false;
+    }
+
+    // We could store an unordered map to speed this up, but with so few items
+    // in the whitelist, the overhead of a map isn't worth it. Iteration is our
+    // primary use-case, and iteration over the vector is fast.
+    for (const auto& project : *m_projects) {
+        if (project.m_name == name) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+WhitelistSnapshot WhitelistSnapshot::Sorted() const
+{
+    // TODO: we cannot sort a vector of const Project objects with std::sort()
+    // because it uses move semantics, so we copy it (twice). Find a better way.
+    std::vector<MutableProject> sorted;
+
+    for (auto const& project : *m_projects) {
+        sorted.emplace_back(project);
+    }
+
+    auto comparer = [] (const MutableProject& a, const MutableProject& b) {
+        return a.LowercaseName() < b.LowercaseName();
+    };
+
+    std::sort(sorted.begin(), sorted.end(), comparer);
+
+    ProjectListPtr copy = std::make_shared<std::vector<Project>>();
+
+    for (auto const& project : sorted) {
+        copy->emplace_back(project.m_name, project.m_url, project.m_timestamp);
+    }
+
+    return WhitelistSnapshot(copy);
+}
+
+// -----------------------------------------------------------------------------
+// Class: Whitelist
+// -----------------------------------------------------------------------------
+
+Whitelist::Whitelist()
+    : m_projects(std::make_shared<std::vector<Project>>())
+{
+}
+
+WhitelistSnapshot Whitelist::Snapshot() const
+{
+    // With C++20, use std::atomic<std::shared_ptr<T>>::load() instead:
+    return WhitelistSnapshot(std::atomic_load(&m_projects));
+}
+
+void Whitelist::Add(
+    const std::string& name,
+    const std::string& url,
+    const int64_t& timestamp)
+{
+    ProjectListPtr copy = std::make_shared<std::vector<Project>>(*m_projects);
```

Comment:
This was by design--see the doc comment on `Whitelist` in project.h. I accepted the potential data race on mutation of the collection as a trade-off to avoid locking because we only write to the whitelist from one thread. As long as that remains true, readers should never see the whitelist in an invalid state. We could beef it up by `std::atomic_load()`-ing the pointer before making a copy. For perfect write-safety, we'd need to change the atomics to locks.

Reply:
The whitelist is only going to be modified from one-thread... so I think this tradeoff is acceptable. @denravonska?

Reply:
Aaaah, gotcha. I had to re-read the code a couple of times but the doc explains it very well.